### ISSUE TITLE
Launch Preparations

### ISSUE THREAD

**Issue by patrickdugan** [OmniLayer/omnicore#182]
I need to do the following:

Verification/rewards program for testers, some copy to put on a page at omnilayer.org/testing

Pull request for level 2 dex. 

Promote page.

Things others need to do:

- [ ] Put big green Download button front-center on Omnilayer.org for windows build.

- [ ] Link to linux instructions.

- [x] Swap GIT link to Omnicore repo not Mastercore.

- [ ] Put me on the team roster. :D

- [ ] Remove people who are no longer active members of the team (optional). 

- [ ] Add an Omnilayer.org/bounties segment where I can then /Futures, /MobileApp ect. and publish the repo requirements documents I have written.

- [ ] Zathras I need you to schedule an AMA which we will promote for the following week (last week of August).

**Comment by dexX7** [comment]
Please provide at least a minimal description what this issue is about?

**Comment by patrickdugan** [comment]
Hey Dexx, we were just talking about these in the meeting, you're good on these, mostly for Adam, Zathras and myself.

**Comment by dexX7** [comment]
Ahh, I see. I just saw the empty entry and was a bit confused.

**Comment by zathras-crypto** [comment]
Sure, though I think it would be much more valuable to do a team AMA to really demonstrate that we don't have individuals as single points of failure & cover a better breadth of topics - please remember I really only focus on OmniChest and OmniCore and so much of the "business" aspects of the project are a bit beyond me.

If @dexX7 would be interested in joining for a developer perspective, along with yourself, Craig for the project perspective, Sean for the OmniJ stuff, Marv/Adam for OmniWallet stuff and so any others I've missed (sorry!) I think that would be much more valuable and provide a much stronger knowledge base.  It could be done over a period of a few days which would provide plenty of time for discussion.


While we're on the topic, I thought I should make sure everyone was really clear that before we can activate ("go live") on the MetaDEx, we need to get final release code into the hands of integrators and users and give them time to upgrade (at least a few weeks since *all* must upgrade) before "go live".  

We only get our release code with testing :)

I only mention it as I thought I heard someone say it would be good to go live on MetaDEx in a week or two at the meeting, which just isn't feasible sadly.

**Comment by dexX7** [comment]
Thanks @zathras-crypto! I'd love to hop it, if there is demand for, and I agree, especially given that Craig was basically exclusively the-answer-guy on Reddit during the last weeks. :) Please ping me via email, if there is anything to discuss/schedule.

> I only mention it as I thought I heard someone say it would be good to go live on MetaDEx in a week or two at the meeting, which just ~~isn't feasible sadly~~.

Actually one week is impossible, and this is enforced by the activation rules, with a minimum grace period of 2048 blocks.

I would really like to see some kind of schedule regarding the activations, given that we currently have three features to activate:

- 1. Class C/OP_RETURN (well tested imho, but very high impact on consensus, and good indicator to see who uses 0.0.10)
- 2. MetaDEx
- 3. Side effects of granting tokens

What comes first? etc.

It should also be noted that we still need release notes, a change log, and ideally some documentation of the configuration and debug options.

**Comment by patrickdugan** [comment]
Ok, we will schedule the AMA for either the last week of August or 1st week of September, will send a mail but you all have to carry the ball with that. 

2048 blocks is 2 weeks. 

My vote would be activate Class C on Sept. 2nd, gives us some testing lead time.

Then it would follow with MetaDex for Sept. 16th.

By side-effects of granting tokens, do you mean for example having a balance of a token that moves another balance to "reserve" (such as the futures contract), can you be specific about what effects?

Dexx can you prepare some of those notes for us? I agree we need to get a simple tutorial page up that let's one get a full Omnicore node up on a server as well.

**Comment by zathras-crypto** [comment]
> Ok, we will schedule the AMA for either the last week of August or 1st week of September, will send a mail but you all have to carry the ball with that. 

No probs re AMA, just let us know when and I'm sure we'll all jump in :)

> 2048 blocks is 2 weeks. 

That's correct, the feature activation system must obey the following notice period on mainnet:
```
    // Notice range for feature activations:
    MIN_ACTIVATION_BLOCKS = 2048;  // ~2 weeks
    MAX_ACTIVATION_BLOCKS = 12288; // ~12 weeks
```

> My vote would be activate Class C on Sept. 2nd, gives us some testing lead time.
> Then it would follow with MetaDex for Sept. 16th.

I hate to be the naysayer on this type of discussion, but managing expectations is critical.  

To activate Class C on Sept 2 means:
* We must bypass the preview & release candidate builds and release current code *immediately* as a full release
* We must ensure all integrators and users (and I do mean *everyone*) is upgraded before Sept 2
* We must send the activation notice 2048 blocks prior to Sept 2 (if we take the two week ballpark, then Aug 19 - roughly two days in the past)

To activate MetaDEx on Sept 16 means:
* We must bypass the preview & release candidate builds and release current code *immediately* as a full release
* Since we would be releasing current code as a full release, this means activating MetaDEx without any of the external testing I've been asking for all these weeks (grossly irresponsible in my view)
* We must ensure all integrators and users (and I do mean *everyone*) is upgraded before Sept 16

To cut a long story short, and I do realize I'm constantly repeating the same thing over and over - but **the hold up for release is the lack of testing on MetaDEx**.  In my eyes the only thing that is going to move us forward at this point is having people who did not build the thing (ie not myself, @dexX7 or @m21) testing it thoroughly and confirming the math.  The math is critical.  Even usability problems can be fixed post activation, but a math problem - especially an exploitable one - well that's a very serious issue.  Due diligence demands we test sufficiently.

Please don't take this the wrong way, I'm keen to move forward too - but not at the expense of risk to our users funds because of a lack of QA.  We even brought on a QA lead at one point, though I didn't get any response to comms with him so have no idea what happened there hehe!!

TL:DR; I will continue to argue against a release until we can build some confidence that we have tested adequately. 

For clarity I have every confidence in our product :) and I'm certainly not trying to be negative, but sometimes engineering must set expectations for the business on what's achievable.   Just trying to be responsible :)

**Comment by dexX7** [comment]
Haha @zathras-crypto, I fully agree with the sentiment, but quick question: does @patrickdugan even know about the activation mechanism? We probably should have introduced it before discussing the timeline ... (unless you did of course).

So just to clarify: features like the MetaDEx, Class C, ... can now remotely be activated by sending a special "activation transaction". The new Omni Core clients will understand the message and the instructions, and the activation could basically sent at any time (in one week, two moths, three years, ...).

However, ideally the activation is broadcasted *after* the users have upgraded, which implies that we must have a finalized release, which is published, announced, distributed and deployed.

Once the activation message is out, there is a software enforced grace period of 2-12 weeks (which we can choose freely), basically as extra fail safe and protection.

It's mandatory that users have up-to-date versions, because otherwise their "view of the system" may be completely different: consider Alice still uses Master Core 0.0.9 without MetaDEx, while Marvin and others use Omni Core 0.0.10, and consider we activate the MetaDEx too soon.

Marvin could then spend all his MSC on the exchange, but since Alice' client doesn't understand the MetaDEx, she would assume Marvin hasn't spent anything. Marvin could then trick Alice, and sell her some MSC for BTC, which he actually doesn't own anymore, but Alice would believe it was a valid trade, and from her point of view, it even looks like a valid trade... until she upgrades and realizes she was scammed.

**Comment by zathras-crypto** [comment]
> Haha @zathras-crypto, I fully agree with the sentiment, but quick question: does @patrickdugan even know about the activation mechanism? We probably should have introduced it before discussing the timeline ... (unless you did of course).

I wouldn't want to speak for @patrickdugan with regards to what he's aware of, but the direction of activation-by-message was set quite a bit back and I did present feature activation at the weekly meeting a couple of weeks ago, and described how you & I had implemented the approach (there were zero questions so perhaps no-one was listening haha) - it's entirely possible @patrickdugan either wasn't there or joined after my section or something though.

I know @CraigSellars and @patrickdugan are very busy, but perhaps we should tee up a meeting with you guys & @dexX7 & I to solidify our next steps (unless @dexX7 you feel like attending the next weekly?)

**Comment by patrickdugan** [comment]
I'm all for that, however next week is non-stop for me. 

However, I did misunderstand the nature of Dexx's question in context. 

We are promoting this, we are starting the re-brand with the Exchanges. 

We will put up 1 formal QA platform and all the community testers we can muster. 

Testing is happening very soon, by Sept. 1 I expect we'll have some decent data.

Given that, we should release 1 RC build for integrators that does both Class C and Meta Dex after at least 1 full-steam week of testing. 

Dexx please clarify what you meant by "side effects", I want to have a clear idea of all specific features/functions as we have them scheduled.

**Comment by dexX7** [comment]
Quick heads up: I'm out of time at the very moment, but I'll answer in detail here, and to your email regarding #80 probably later. :)

**Comment by dexX7** [comment]
> @zathras-crypto: unless @dexX7 you feel like attending the next weekly?

Sounds fine, let's do it, however it looks like this doesn't work out for @patrickdugan, as far as I can see.

> @patrickdugan: Dexx please clarify what you meant by "side effects", I want to have a clear idea of all specific features/functions as we have them scheduled.

Are you familiar with a) Class C/OP_RETURN and b) the MetaDEx?

When currently granting tokens for a [managed property via tx 55](https://github.com/OmniLayer/spec#granting-tokens-for-a-managed-property), and the recipient of the tokens has an active crowdsale that accepts these tokens, then this triggers the effect of "crowdsale participation", similar as if one would have [participated in the crowdsale via simple send](https://github.com/OmniLayer/spec#participating-in-a-crowdsale).

However, it was, to my knowledge, never intended, nor specified this way, and it was simply a side effect of the underlying implementation.

We believe that actions should be active and explicit, instead of implicit, and decided to propose the deactivation of the passive crowdsale participation in this case.

This doesn't limit the potential actions of the protocol, because the recipient of the granted tokens may still participate in the crowdsale, but it would require an explicit action, namely a simple send, to trigger the effect.

I stumbled upon this almost a year ago in a testing environment, and so far it was never actually observed in the wild on mainnet, which underlines that the deactivation likely won't annoy or affect anyone.

If the feature with identifier 4 is activated, the side effect will be deactivated, and granting tokens will no longer trigger passive crowdsale participations.

The related pull request and related issues on this are #163, #157 and https://github.com/mastercoin-MSC/mastercore/issues/227.

**Comment by patrickdugan** [comment]
Hey I put in probably 35 or 40 hours this week just on Omni Foundation stuff so I ought to make the meeting Tuesday right? I am finishing a speech 15 minutes before the appointed time and will be perhaps a bit late.

Regarding this side effect you described, reminds me of the issue of having algorithmically managed tokens that attract dividends. There's some risk that you sprinkle dividends in algo-controlled addresses from time to time. Edge cases but ones that should be thought through, in that case I suppose having a check for any dividends and sweeping those balances into the simple send batch of whatever the algo does next (could also apply to limit order escrow addresses). Or a "send-all" flag associted with escrow outcomes.

**Comment by dexX7** [comment]
> I ought to make the meeting Tuesday right? I am finishing a speech 15 minutes before the appointed time and will be perhaps a bit late.

Works for me, and I don't mind delays, if there are any.

> Regarding this side effect you described ...

The effect to be disabled is very specific and likely rare: there would need to be a managed property A, and a crowdsale B, which accepts A tokens, and further the issuer of A would need to grant new A tokens, and choose the address of crowdsale B as destination.

If such an effect is desired for an use-case you're currently working on, we'd have to reevaluate, whether we actually should disable it. However, the huge upside of the whole activation feature: we don't *have* to use it, but we *can*, i.e. this doesn't affect, whether 0.0.10 can be released or not, and it would be fine, if we never activate the feature (thus never disable the side effect) at all.

**Comment by dexX7** [comment]
Regarding this issue here, I'm inclined to close it in favor of #194, however, we may also continue to use it for general discussion about the release. Please let me know what you think, @patrickdugan.

**Comment by zathras-crypto** [comment]
>>  I ought to make the meeting Tuesday right? I am finishing a speech 15 minutes before the appointed time and will be perhaps a bit late.
> Works for me, and I don't mind delays, if there are any.

Awesome, do you have an invite to sococo web?  Details on timing etc? I'll suggest we depart from the usual and head into a thorough discussion on how to progress things and roadmap the next few weeks :)

**Comment by dexX7** [comment]
Hehe, please don't get too excited. With the setup here I'm able to listen and type, and in the past Sococo was quite a PITA for me, due to audio/driver issues (finally resolved though! :)).

It's 21:00 UTC, right? Would be great, if we can fill #194.

**Comment by zathras-crypto** [comment]
> It's 21:00 UTC, right?

Yep.

> With the setup...

No probs, text is fine :)

**Comment by dexX7** [comment]
For the sake of #194 I'm going to close this. Please reopen, if you like to leave this issue open.
